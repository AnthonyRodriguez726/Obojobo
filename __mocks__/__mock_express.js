// This file is used to set up Mocks for Express Methods and
// Router Methods. It is used primarily in __tests___/routes

let mockExpressMethods = {}
let mockRouterMethods = {}

let mockExpress = () => {
	jest.mock(
		'express',
		() => {
			let module = () => {
				let methods = ['on', 'use', 'get', 'post', 'put', 'delete', 'all', 'static']
				let obj = {}
				methods.forEach(m => {
					obj[m] = mockExpressMethods[m] = jest.fn()
				})
				return obj
			}

			module.Router = () => {
				// create all the router verbs as jest functions
				let theVerbs = ['all', 'get', 'post', 'delete', 'put']
				let mockVerbs = {}
				theVerbs.forEach(m => {
					mockVerbs[m] = mockRouterMethods[m] = jest.fn().mockReturnValue(mockVerbs)
				})
				mockRouterMethods.route = mockVerbs.route = jest.fn().mockReturnValue(mockVerbs)

				// When route('/website').post() is called
				// we're going to want to know the context for the post
				// so we'll keep the main argument for route in scope
				// obj.route = jest.fn().mockImpleme(theRoute => {

				return mockVerbs
			}

			return module
		},
		{ virtual: true }
	)
}

mockExpress()

module.exports = {
	mockExpressMethods,
	mockRouterMethods
}
