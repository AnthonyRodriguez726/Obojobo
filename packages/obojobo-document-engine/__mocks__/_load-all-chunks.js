import '../ObojoboDraft/Chunks/ActionButton/viewer'
import '../ObojoboDraft/Chunks/Break/viewer'
import '../ObojoboDraft/Chunks/Code/viewer'
import '../ObojoboDraft/Chunks/Figure/viewer'
import '../ObojoboDraft/Chunks/Heading/viewer'
import '../ObojoboDraft/Chunks/HTML/viewer'
import '../ObojoboDraft/Chunks/IFrame/viewer'
import '../ObojoboDraft/Chunks/List/viewer'
import '../ObojoboDraft/Chunks/MathEquation/viewer'
import '../ObojoboDraft/Chunks/MCInteraction/viewer'
import '../ObojoboDraft/Chunks/Question/viewer'
import '../ObojoboDraft/Chunks/QuestionBank/viewer'
import '../ObojoboDraft/Chunks/Table/viewer'
import '../ObojoboDraft/Chunks/Text/viewer'
import '../ObojoboDraft/Chunks/YouTube/viewer'
import '../ObojoboDraft/Pages/Page/viewer'
import '../ObojoboDraft/Sections/Content/viewer'
import '../ObojoboDraft/Sections/Assessment/viewer'
import '../ObojoboDraft/Modules/Module/viewer'
