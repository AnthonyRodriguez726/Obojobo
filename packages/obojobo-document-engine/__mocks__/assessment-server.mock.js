let getAttemptStartServerResponse = () => {
	return {
		status: 'ok',
		value: {
			assessmentId: 'assessment',
			attemptId: '00000000-0000-0000-0000-000000000000',
			endTime: null,
			startTime: '1970-01-01T00:00:00.000Z',
			state: {
				data: {},
				qb: {
					id: 'c8587164-f7a2-4927-9548-58b8a94b86c9',
					type: 'ObojoboDraft.Chunks.QuestionBank',
					content: {
						choose: 1,
						select: 'sequential'
					},
					children: [
						{
							id: 'qb1',
							type: 'ObojoboDraft.Chunks.QuestionBank',
							content: {
								choose: 1,
								select: 'sequential'
							},
							children: [
								{
									id: 'qb1-q1',
									type: 'ObojoboDraft.Chunks.Question',
									content: {},
									children: [
										{
											id: '8dffa3d6-9185-46a0-9b0c-bc2393532ccf',
											type: 'ObojoboDraft.Chunks.Text',
											content: {
												textGroup: [
													{
														text: {
															value:
																"Question Bank 1 Question 1 - This is the first question you'll see",
															styleList: []
														},
														data: null
													}
												]
											},
											children: []
										},
										{
											id: 'qb1-q1-mca',
											type: 'ObojoboDraft.Chunks.MCInteraction',
											content: {
												responseType: 'pick-one'
											},
											children: [
												{
													id: 'qb1-q1-mca-mc1',
													type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
													content: {
														score: 100
													},
													children: [
														{
															id: 'qb1-q1-mca-mc1-ans',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
															content: {},
															children: [
																{
																	id: '9bb1c551-48be-4406-9cdb-fa21044168ca',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Correct',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														},
														{
															id: 'qb1-q1-mca-mc1-fb',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCFeedback',
															content: {},
															children: [
																{
																	id: 'qb1-q1-mca-mc1-fb-t',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Correct',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														}
													]
												},
												{
													id: 'qb1-q1-mca-mc2',
													type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
													content: {
														score: 0
													},
													children: [
														{
															id: 'qb1-q1-mca-mc2-ans',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
															content: {},
															children: [
																{
																	id: '6f271e7e-15ad-4467-9ea6-66ebf625c4fa',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Incorrect',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														},
														{
															id: 'qb1-q1-mca-mc2-fb',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCFeedback',
															content: {},
															children: [
																{
																	id: 'qb1-q1-mca-mc2-fb-t',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Incorrect',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														}
													]
												}
											]
										}
									]
								},
								{
									id: 'qb1-q2',
									type: 'ObojoboDraft.Chunks.Question',
									content: {},
									children: [
										{
											id: '507d21a8-7ae6-4978-9343-6de0923bb0bc',
											type: 'ObojoboDraft.Chunks.Text',
											content: {
												textGroup: [
													{
														text: {
															value:
																"Question Bank 1 Question 2 - This is the fourth question you'll see",
															styleList: []
														},
														data: null
													}
												]
											},
											children: []
										},
										{
											id: 'qb1-q2-mca',
											type: 'ObojoboDraft.Chunks.MCInteraction',
											content: {
												responseType: 'pick-one'
											},
											children: [
												{
													id: 'qb1-q2-mca-mc1',
													type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
													content: {
														score: 100
													},
													children: [
														{
															id: 'qb1-q2-mca-mc1-ans',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
															content: {},
															children: [
																{
																	id: 'a4f51a79-fb8c-4cff-86fa-3da3423cfe77',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Correct',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														}
													]
												},
												{
													id: 'qb1-q2-mca-mc2',
													type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
													content: {
														score: 0
													},
													children: [
														{
															id: 'qb1-q2-mca-mc2-ans',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
															content: {},
															children: [
																{
																	id: 'c5821823-9a39-4f78-ae20-71de2baad1ed',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Incorrect',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														}
													]
												}
											]
										}
									]
								}
							]
						},
						{
							id: 'qb2',
							type: 'ObojoboDraft.Chunks.QuestionBank',
							content: {},
							children: [
								{
									id: 'qb2.q1',
									type: 'ObojoboDraft.Chunks.Question',
									content: {},
									children: [
										{
											id: 'ef787e6c-a7cf-40a5-9b34-4507b955326f',
											type: 'ObojoboDraft.Chunks.Text',
											content: {
												textGroup: [
													{
														text: {
															value:
																"Question Bank 2 Question 1 - This is the second question you'll see",
															styleList: []
														},
														data: null
													}
												]
											},
											children: []
										},
										{
											id: 'qb2.q1-mca',
											type: 'ObojoboDraft.Chunks.MCInteraction',
											content: {
												responseType: 'pick-one'
											},
											children: [
												{
													id: 'qb2.q1-mca-mc1',
													type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
													content: {
														score: 100
													},
													children: [
														{
															id: 'qb2.q1-mca-mc1-ans',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
															content: {},
															children: [
																{
																	id: '7207dfd6-a171-4fe9-8a1a-b2614fdb9873',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Correct',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														}
													]
												},
												{
													id: 'qb2.q1-mca-mc2',
													type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
													content: {
														score: 0
													},
													children: [
														{
															id: 'qb2.q1-mca-mc2-ans',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
															content: {},
															children: [
																{
																	id: 'e3829d20-f884-4e5a-96d7-7b2eac59ee4d',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Incorrect',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														}
													]
												}
											]
										}
									]
								},
								{
									id: 'qb2.q2',
									type: 'ObojoboDraft.Chunks.Question',
									content: {},
									children: [
										{
											id: '2bda57b5-254e-4465-8572-23d5250cc715',
											type: 'ObojoboDraft.Chunks.Text',
											content: {
												textGroup: [
													{
														text: {
															value:
																"Question Bank 2 Question 2 - This is the third question you'll see",
															styleList: []
														},
														data: null
													}
												]
											},
											children: []
										},
										{
											id: 'qb2.q2-mca',
											type: 'ObojoboDraft.Chunks.MCInteraction',
											content: {
												responseType: 'pick-one'
											},
											children: [
												{
													id: 'qb2.q2-mca-mc1',
													type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
													content: {
														score: 100
													},
													children: [
														{
															id: 'qb2.q2-mca-mc1-ans',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
															content: {},
															children: [
																{
																	id: '12d8347b-95e1-4324-9d32-24c68d318ecb',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Correct',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														}
													]
												},
												{
													id: 'qb2.q2-mca-mc2',
													type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
													content: {
														score: 0
													},
													children: [
														{
															id: 'qb2.q2-mca-mc2-ans',
															type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
															content: {},
															children: [
																{
																	id: '02ed7f55-8181-4431-9d35-96dbb778221a',
																	type: 'ObojoboDraft.Chunks.Text',
																	content: {
																		textGroup: [
																			{
																				text: {
																					value: 'Incorrect',
																					styleList: []
																				},
																				data: null
																			}
																		]
																	},
																	children: []
																}
															]
														}
													]
												}
											]
										}
									]
								}
							]
						}
					]
				},
				questions: [
					{
						id: 'qb1-q1',
						type: 'ObojoboDraft.Chunks.Question',
						content: {},
						children: [
							{
								id: '8dffa3d6-9185-46a0-9b0c-bc2393532ccf',
								type: 'ObojoboDraft.Chunks.Text',
								content: {
									textGroup: [
										{
											text: {
												value: "Question Bank 1 Question 1 - This is the first question you'll see",
												styleList: []
											},
											data: null
										}
									]
								},
								children: []
							},
							{
								id: 'qb1-q1-mca',
								type: 'ObojoboDraft.Chunks.MCInteraction',
								content: {
									responseType: 'pick-one'
								},
								children: [
									{
										id: 'qb1-q1-mca-mc1',
										type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
										content: {
											score: 100
										},
										children: [
											{
												id: 'qb1-q1-mca-mc1-ans',
												type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
												content: {},
												children: [
													{
														id: '9bb1c551-48be-4406-9cdb-fa21044168ca',
														type: 'ObojoboDraft.Chunks.Text',
														content: {
															textGroup: [
																{
																	text: {
																		value: 'Correct',
																		styleList: []
																	},
																	data: null
																}
															]
														},
														children: []
													}
												]
											}
										]
									},
									{
										id: 'qb1-q1-mca-mc2',
										type: 'ObojoboDraft.Chunks.MCInteraction.MCChoice',
										content: {
											score: 0
										},
										children: [
											{
												id: 'qb1-q1-mca-mc2-ans',
												type: 'ObojoboDraft.Chunks.MCInteraction.MCAnswer',
												content: {},
												children: [
													{
														id: '6f271e7e-15ad-4467-9ea6-66ebf625c4fa',
														type: 'ObojoboDraft.Chunks.Text',
														content: {
															textGroup: [
																{
																	text: {
																		value: 'Incorrect',
																		styleList: []
																	},
																	data: null
																}
															]
														},
														children: []
													}
												]
											},
											{
												id: 'qb1-q1-mca-mc2-fb',
												type: 'ObojoboDraft.Chunks.MCInteraction.MCFeedback',
												content: {},
												children: [
													{
														id: 'qb1-q1-mca-mc2-fb-t',
														type: 'ObojoboDraft.Chunks.Text',
														content: {
															textGroup: [
																{
																	text: {
																		value: 'Incorrect',
																		styleList: []
																	},
																	data: null
																}
															]
														},
														children: []
													}
												]
											}
										]
									}
								]
							}
						]
					}
				]
			}
		}
	}
}

let getAttemptEndServerResponse = questionScore => {
	let startResp = getAttemptStartServerResponse()

	return {
		status: 'ok',
		value: {
			assessmentId: 'assessment',
			attemptId: '00000000-0000-0000-0000-000000000000',
			startTime: '1970-01-01T00:00:00.000Z',
			endTime: '1970-01-01T00:00:10.000Z',
			ltiOutcomes: {
				sent: false
			},
			result: {
				attemptScore: questionScore,
				scores: [{ id: 'qb1-q1', score: questionScore }]
			},
			state: startResp.state
		}
	}
}

export { getAttemptStartServerResponse, getAttemptEndServerResponse }
