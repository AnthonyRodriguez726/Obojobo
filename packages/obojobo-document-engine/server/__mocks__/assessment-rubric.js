const mockGetAssessmentScoreInfoForAttempt = jest.fn()

class AssessmentRubric {
	constructor(rubric) {
		this.getAssessmentScoreInfoForAttempt = mockGetAssessmentScoreInfoForAttempt
	}
}

AssessmentRubric.mockGetAssessmentScoreInfoForAttempt = mockGetAssessmentScoreInfoForAttempt

module.exports = AssessmentRubric
