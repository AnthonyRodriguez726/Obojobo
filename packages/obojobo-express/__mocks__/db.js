let db = jest.fn()

db.one = jest.fn()
db.one.mockImplementation((query, vars) => {
	return Promise.resolve(null)
})

db.any = jest.fn()
db.any.mockImplementation((query, vars) => {
	return Promise.resolve(null)
})

db.none = jest.fn()
db.none.mockImplementation((query, vars) => {
	return Promise.resolve(null)
})

module.exports = db;
