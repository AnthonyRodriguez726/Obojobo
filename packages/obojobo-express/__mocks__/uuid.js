module.exports = {
	v4: jest.fn(() => 'fb9826d0-3d7a-43b8-a5ad-3949ce620d91')
}
