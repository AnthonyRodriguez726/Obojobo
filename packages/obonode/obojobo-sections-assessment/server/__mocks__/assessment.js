const Assessment = {
	completeAttempt: jest.fn(),
	getAttempt: jest.fn().mockResolvedValue({
		id: 'mockAttemptId',
		user_id: 'mockUserId',
		draft_id: 'mockDraftId',
		assessment_id: 'mockAssessmentId',
		state: 'mockState'
	}),
	getAttempts: jest.fn(),
	getAttemptNumber: jest.fn().mockResolvedValue('mockAttemptNumber'),
	getCompletedAssessmentAttemptHistory: jest.fn(),
	getResponsesForAttempt: jest.fn()
}

module.exports = Assessment
