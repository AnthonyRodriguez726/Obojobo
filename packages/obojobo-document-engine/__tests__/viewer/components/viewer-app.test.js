import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme'
import OboModel from '../../../__mocks__/_obo-model-with-chunks'
import {
	getAttemptStartServerResponse,
	getAttemptEndServerResponse
} from '../../../__mocks__/assessment-server.mock'
import {
	AssessmentStore,
	NavStore,
	QuestionStore,
	ScoreStore,
	ModalStore,
	FocusStore,
	AssessmentUtil,
	NavUtil,
	QuestionUtil,
	ScoreUtil,
	ModalUtil,
	FocusUtil
} from '../../../__mocks__/viewer-state.mock'
import Dispatcher from '../../../src/scripts/common/flux/dispatcher'
import ViewerApp from '../../../src/scripts/viewer/components/viewer-app'
// import OboModel from '../../../__mocks__/_obo-model-with-chunks'
import '../../../__mocks__/_load-all-chunks'

import APIUtil from '../../../src/scripts/viewer/util/api-util'
// console.log('moduleData be all', moduleData.type)

jest.mock('../../../src/scripts/viewer/util/api-util', () => {
	return {
		get: jest.fn(),
		post: jest.fn(),
		postEvent: jest.fn(),
		saveState: jest.fn(),
		fetchDraft: jest.fn(),
		getAttempts: jest.fn(),
		startAttempt: jest.fn(),
		endAttempt: jest.fn()
	}
})

APIUtil.postEvent = () => {
	return Promise.resolve({ status: 'ok' })
}
APIUtil.startAttempt = () => {
	return Promise.resolve(getAttemptStartServerResponse())
}
APIUtil.endAttempt = () => {
	return Promise.resolve(getAttemptEndServerResponse(100))
}

describe('ViewerApp', () => {
	let json = require('../../../test-object.json')
	// let viewerEl;

	beforeEach(() => {
		jest.clearAllMocks()

		window.__oboGlobals.draft = json
		window.__oboGlobals.previewing = 'false'
		window.__oboGlobals['ObojoboDraft.Sections.Assessment:attemptHistory'] = []

		// viewerEl = mount(<ViewerApp />)
	})

	test("Navigation doesn't change when navigation is locked but does when unlocked", () => {
		let viewerEl = mount(<ViewerApp />)

		NavUtil.lock()
		let html1 = viewerEl.html()

		NavUtil.goPrev()
		let html2 = viewerEl.html()

		NavUtil.unlock()
		let html3 = viewerEl.html()

		NavUtil.goNext()
		let html4 = viewerEl.html()

		expect(html1).toEqual(html2)
		expect(html3).not.toEqual(html4)

		viewerEl.unmount()
	})

	test("Navigation doesn't navigate to a page that doesn't exist", () => {
		let viewerEl = mount(<ViewerApp />)

		NavUtil.unlock()
		let html1 = viewerEl.html()

		NavUtil.goPrev()
		let html2 = viewerEl.html()

		NavUtil.goto('assessment') // last page
		let html3 = viewerEl.html()

		NavUtil.goNext()
		let html4 = viewerEl.html()

		expect(html1).toEqual(html2)
		expect(html3).toEqual(html4)

		viewerEl.unmount()
	})

	test("Navigation doesn't navigate to a page that doesn't exist", () => {
		let viewerEl = mount(<ViewerApp />)

		NavUtil.unlock()
		let html1 = viewerEl.html()

		NavUtil.goPrev()
		let html2 = viewerEl.html()

		NavUtil.goto('assessment') // last page
		let html3 = viewerEl.html()

		NavUtil.goNext()
		let html4 = viewerEl.html()

		expect(html1).toEqual(html2)
		expect(html3).toEqual(html4)

		viewerEl.unmount()
	})

	test('Prev/Next buttons go to prev/next page', () => {
		let viewerEl = mount(<ViewerApp />)
		let prevBtnEl = viewerEl.find('.viewer--components--inline-nav-button.is-prev')
		let nextBtnEl = viewerEl.find('.viewer--components--inline-nav-button.is-next')

		expect(viewerEl.find('#obo-page-1').length).toBe(1)

		// click back, should do nothing
		prevBtnEl.simulate('click')
		expect(viewerEl.find('#obo-page-1').length).toBe(1)

		// click forward, should move forward
		nextBtnEl.simulate('click')
		expect(viewerEl.find('#obo-page-2').length).toBe(1)

		// click back, should go back
		prevBtnEl.simulate('click')
		expect(viewerEl.find('#obo-page-1').length).toBe(1)

		// go to the final page
		nextBtnEl.simulate('click')
		nextBtnEl.simulate('click')
		nextBtnEl.simulate('click')
		nextBtnEl.simulate('click')
		nextBtnEl.simulate('click')
		nextBtnEl.simulate('click')
		expect(viewerEl.find('#obo-assessment').length).toBe(1)

		// click next, should do nothing
		nextBtnEl.simulate('click')
		expect(viewerEl.find('#obo-assessment').length).toBe(1)

		viewerEl.unmount()
	})

	test('Clicking on a question shows it', () => {
		let viewerEl = mount(<ViewerApp />)

		NavUtil.goto('page-3')

		let questionEl = viewerEl.find('#obo-pq1')

		expect(QuestionStore.getState().viewedQuestions).toEqual({})
		expect(QuestionStore.getState().viewing).toBe(null)

		questionEl.find('.blocker').simulate('click')

		expect(QuestionStore.getState().viewedQuestions).toEqual({
			pq1: true
		})
		expect(QuestionStore.getState().viewing).toBe('pq1')

		viewerEl.unmount()
	})

	test("Answering a question incorrectly displays 'Incorrect' and shows feedback", () => {
		let viewerEl = mount(<ViewerApp />)

		NavUtil.goto('page-3')

		let questionEl = viewerEl.find('#obo-pq1')

		expect(QuestionStore.getState().responses).toEqual({})
		expect(questionEl.find('.result.correct').length).toBe(0) // Correct label
		expect(questionEl.find('.result.incorrect').length).toBe(0) // Incorrect label
		expect(questionEl.find('#obo-pq1-mca-mc3-fb').length).toBe(0) // Feedback
		expect(questionEl.find('button').length).toBe(2) // No 'Solution' button

		questionEl.find('.blocker').simulate('click')
		questionEl.find('#obo-pq1-mca-mc3').simulate('click') // wrong answer choice
		questionEl.find('.submit button').simulate('click') // Check your answer button

		expect(questionEl.find('.result.correct').length).toBe(0)
		expect(questionEl.find('.result.incorrect').length).toBe(1)
		expect(questionEl.find('#obo-pq1-mca-mc3-fb').length).toBe(1)
		expect(questionEl.find('button').length).toBe(3) // 'Solution' button

		expect(QuestionStore.getState().responses).toEqual({
			pq1: { ids: ['pq1-mca-mc3'] }
		})

		let scores = ScoreStore.getState().scores
		expect(scores.pq1).toBeDefined()
		expect(Object.keys(scores.pq1).sort()).toEqual(['id', 'itemId', 'score'])
		expect(scores.pq1.itemId).toEqual('pq1')
		expect(scores.pq1.score).toEqual(0)

		viewerEl.unmount()
	})

	test("Answering a question correctly displays 'Correct' and doesn't show feedback if none exists", () => {
		let viewerEl = mount(<ViewerApp />)

		NavUtil.goto('page-3')

		let questionEl = viewerEl.find('#obo-pq1')

		expect(QuestionStore.getState().responses).toEqual({})
		expect(questionEl.find('.result.correct').length).toBe(0) // Correct label
		expect(questionEl.find('.result.incorrect').length).toBe(0) // Incorrect label
		expect(questionEl.find('button').length).toBe(2) // No 'Solution' button

		questionEl.find('.blocker').simulate('click')
		questionEl.find('#obo-pq1-mca-mc1').simulate('click') // correct answer choice
		questionEl.find('.submit button').simulate('click') // Check your answer button

		expect(questionEl.find('.result.correct').length).toBe(1)
		expect(questionEl.find('.result.incorrect').length).toBe(0)
		expect(questionEl.find('.solution .score').textContent).toBe(undefined)
		expect(questionEl.find('button').length).toBe(3) // 'Solution' button

		expect(QuestionStore.getState().responses).toEqual({
			pq1: { ids: ['pq1-mca-mc1'] }
		})

		let scores = ScoreStore.getState().scores
		expect(scores.pq1).toBeDefined()
		expect(Object.keys(scores.pq1).sort()).toEqual(['id', 'itemId', 'score'])
		expect(scores.pq1.itemId).toEqual('pq1')
		expect(scores.pq1.score).toEqual(100)

		viewerEl.unmount()
	})

	test('Clicking the button to show the solution will show the solution', () => {
		let viewerEl = mount(<ViewerApp />)

		NavUtil.goto('page-3')

		let questionEl = viewerEl.find('#obo-pq1')

		expect(questionEl.find('.solution-container').length).toBe(0)

		questionEl.find('.blocker').simulate('click')
		questionEl.find('#obo-pq1-mca-mc3').simulate('click') // wrong answer choice
		questionEl.find('.submit button').simulate('click') // Check your answer button
		// expect(questionEl.find('button').at(1)).toBe(1)
		questionEl.find('button').at(1).simulate('click') // Show solution button

		expect(questionEl.find('.solution-container').length).toBe(1)

		viewerEl.unmount()
	})

	test('Clicking start assessment will lock out navigation and start the assessment', done => {
		let onAttemptStarted = () => {
			Dispatcher.off('assessment:attemptStarted', onAttemptStarted)

			expect(viewerEl.find('#obo-assessment').length).toBe(1)
			expect(viewerEl.find('#obo-page-1').length).toBe(0)

			// preview mode allows us to unlock the navigation however
			let previewBannerEl = viewerEl.find('.preview-banner')
			let unlockButtonEl = previewBannerEl.find('button').at(0)

			unlockButtonEl.simulate('click')
			firstPgLinkEl.simulate('click')

			expect(viewerEl.find('#obo-assessment').length).toBe(0)
			expect(viewerEl.find('#obo-page-1').length).toBe(1)

			viewerEl.unmount()

			done()
		}
		let viewerEl = mount(<ViewerApp />)

		NavUtil.goto('assessment')

		let buttonEl = viewerEl.find('#obo-start-assessment-button')

		buttonEl.find('button').simulate('click')

		expect(viewerEl.find('#obo-assessment').length).toBe(1)
		expect(viewerEl.find('#obo-page-1').length).toBe(0)

		let navEl = viewerEl.find('.viewer--components--nav')
		let firstPgLinkEl = navEl.find('a').at(1)

		Dispatcher.on('assessment:attemptStarted', onAttemptStarted)
		firstPgLinkEl.simulate('click')
	})

	test('Finishing an assessment shows a score', done => {
		let onAttemptStarted = () => {
			Dispatcher.off('assessment:attemptStarted', onAttemptStarted)

			let q1 = viewerEl.find('#obo-qb1-q1')
			q1.find('.flipper').simulate('click')

			let c1 = viewerEl.find('#obo-qb1-q1-mca-mc1')

			c1.simulate('click')

			viewerEl.find('.submit-button').find('button').simulate('click')
		}

		let onAttemptEnded = () => {
			Dispatcher.off('assessment:attemptEnded', onAttemptEnded)

			let assessmentEl = viewerEl.find('#obo-assessment')
			let score = assessmentEl.find('.score')
			let assessment = AssessmentStore.getState().assessments.assessment
			let attempt = assessment.attempts[0]

			expect(assessment.attempts.length).toBe(1)
			expect(attempt.result).toEqual({
				attemptScore: 100,
				scores: [
					{
						id: 'qb1-q1',
						score: 100
					}
				]
			})

			viewerEl.unmount()

			done()
		}

		let viewerEl = mount(<ViewerApp />)

		NavUtil.goto('assessment')

		Dispatcher.on('assessment:attemptStarted', onAttemptStarted)
		Dispatcher.on('assessment:attemptEnded', onAttemptEnded)

		let buttonEl = viewerEl.find('#obo-start-assessment-button')
		buttonEl.find('button').simulate('click')
	})
})
