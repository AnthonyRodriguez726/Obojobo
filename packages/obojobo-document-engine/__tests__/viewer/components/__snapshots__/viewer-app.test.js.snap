// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewerApp clearPreviewScores shows error dialog 1`] = `
Array [
  "modal:show",
  Object {
    "value": Object {
      "component": <SimpleDialog
        cancelOk={false}
        noOrYes={false}
        ok={true}
        onCancel={[Function]}
        onConfirm={[Function]}
        title={null}
        width="15em"
        yesOrNo={false}
      >
        There was an error resetting assessments and questions
      </SimpleDialog>,
      "hideViewer": false,
    },
  },
]
`;
